cpp/ql/src/experimental/Security/CWE/CWE-476/DangerousUseOfExceptionBlocks.ql

```diff
@@ -0,0 +1,165 @@
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-476
+ *       external/cwe/cwe-415
+ */
+
+import cpp
+
+/** Holds if the release can occur twice. in the current block of catch and above in the block of try or other block catch. */
+pragma[inline]
+predicate doubleCallDelete(CatchAnyBlock cb, Variable vr) {
+  // Search for exceptions after freeing memory.
+  exists(Expr e1 |
+    (
+      e1 = vr.getAnAccess().getEnclosingStmt().(ExprStmt).getExpr().(DeleteArrayExpr) or
+      e1 = vr.getAnAccess().getEnclosingStmt().(ExprStmt).getExpr().(DeleteExpr)
+    ) and
+    e1.getEnclosingFunction() = cb.getEnclosingFunction() and
+    (
+      e1.getEnclosingStmt().getParentStmt*() = cb.getTryStmt().getStmt() and
+      exists(Expr e2, ThrowExpr th |
+        (
+          e2 = th or
+          e2 = th.getEnclosingFunction().getACallToThisFunction()
+        ) and
+        e2.getEnclosingStmt().getParentStmt*() = cb.getTryStmt().getStmt() and
+        e1.getASuccessor+() = e2
+      ) and
+      not exists(AssignExpr ae |
+        ae.getLValue().(VariableAccess).getTarget() = vr and
+        ae.getRValue().getValue() = "0" and
+        e1.getASuccessor+() = ae and
+        ae.getEnclosingStmt().getParentStmt*() = cb.getTryStmt().getStmt()
+      )
+      or
+      // Search for a situation when there is a higher catch block that also frees memory.
+      exists(CatchBlock cbt, Expr e2, ThrowExpr th |
+        e1.getEnclosingStmt().getParentStmt*() = cbt and
+        exists(cbt.getParameter()) and
+        (
+          e2 = th or
+          e2 = th.getEnclosingFunction().getACallToThisFunction()
+        ) and
+        e2.getEnclosingStmt().getParentStmt*() = cbt and
+        e1.getASuccessor+() = e2 and
+        not exists(AssignExpr ae |
+          ae.getLValue().(VariableAccess).getTarget() = vr and
+          ae.getRValue().getValue() = "0" and
+          e1.getASuccessor+() = ae and
+          ae.getEnclosingStmt().getParentStmt*() = cbt
+        )
+      )
+    ) and
+    // Exclude the presence of a check in catch block.
+    not exists(IfStmt ifst |
+      ifst.getEnclosingStmt().getParentStmt*() = cb.getAStmt()
+    )
+  )
+}
+
+/** Holds if an exception can be thrown before the memory is allocated, and when the exception is handled, an attempt is made to access unallocated memory in the catch block. */
+pragma[inline]
+predicate pointerDereference(CatchAnyBlock cb, Variable vr, Variable vro) {
+  // Search exceptions before allocating memory.
+  exists(Expr e0, Expr e1 |
+    (
+      exists(AssignExpr ase |
+        ase = vro.getAnAccess().getEnclosingStmt().(ExprStmt).getExpr().(AssignExpr) and
+        (
+          e0 = ase.getRValue().(NewOrNewArrayExpr) or
+          e0 = ase.getRValue().(NewOrNewArrayExpr).getEnclosingFunction().getACallToThisFunction()
+        ) and
+        vro = ase.getLValue().(VariableAccess).getTarget()
+      )
+      or
+      exists(AssignExpr ase |
```

Comment:
I'm not exactly sure what this part is for, but please could you add a comment explaining it?

Reply:
this is necessary to account for indirect access to a variable, e.g.
```cpp
vro[i] = new vroType;
pointVro->vro[i] = new vroType;
pointVro.vro[i] = new vroType;
```
do you think the following comment is acceptable
// `e0` is a `new expression` (or equivalent function call) assigned to the array element `vro`

Reply:
Yes I think so.  The more chunks of code that have comments, the easier it is to understand what's going on.

Reply:
:+1: